denops/ddc/ddc.ts

```diff
@@ -125,82 +169,105 @@ export class Ddc {
     denops: Denops,
     context: Context,
     options: DdcOptions,
-  ): Promise<[number, DdcCandidate[]]> {
-    let completePos = -1;
-    let candidates: DdcCandidate[] = [];
-    const sources = options.sources.map((name) => this.sources[name])
-      .filter((x) => x);
-
-    for (const source of sources) {
-      const [sourceOptions, sourceParams] = sourceArgs(options, source);
-      completePos = await source.getCompletePosition(
-        denops,
-        context,
-        sourceOptions,
-        sourceParams,
-      );
-      const sourceCandidates = await source.gatherCandidates(
-        denops,
-        context,
-        sourceOptions,
-        sourceParams,
-      );
-      const filterCandidates = await this.filterCandidates(
-        denops,
-        context,
-        sourceOptions,
-        options.filterOptions,
-        options.filterParams,
-        sourceCandidates,
-      );
-      const result: DdcCandidate[] = filterCandidates.map((c: Candidate) => (
-        {
-          ...c,
-          abbr: formatAbbr(c.word, c.abbr),
-          source: source.name,
-          icase: true,
-          equal: true,
-          menu: formatMenu(sourceOptions.mark, c.menu),
-        }
-      ));
-
-      candidates = candidates.concat(result);
-    }
-
-    return [completePos, candidates];
+  ): Promise<
+    { completePos: number; candidates: ReadableStream<DdcCandidate[]> }
+  > {
+    const sources = await Promise.all(
+      options.sources
+        .map((n) => this.sources[n])
+        .filter((v) => v)
+        .map(async (s) => {
+          const [sourceOptions, sourceParams] = sourceArgs(options, s);
+          const [completePos, candidates] = await Promise.all([
+            s.getCompletePosition(
+              denops,
+              context,
+              sourceOptions,
+              sourceParams,
+            ),
+            s.gatherCandidates(
+              denops,
+              context,
+              sourceOptions,
+              sourceParams,
+            ),
+          ]);
+          return {
+            completePos,
+            candidates: candidates
+              .pipeThrough(this.filterCandidates(
+                denops,
+                context,
+                sourceOptions,
+                options.filterOptions,
+                options.filterParams,
+              ))
+              .pipeThrough(this.finalizeCandidates(
+                s,
+                sourceOptions,
+              )),
+          };
+        }),
+    );
+    const completePos = Math.max(
+      ...sources.map(({ completePos }) => completePos),
+    );
```

Comment:
I've no idea what is the `completePos` thus I just use the biggest one (but I'm sure it's wrong)

Reply:
`completePos` means source complete position.
It should be merged by smallest one.